src/internal/page-key.ts

```diff
@@ -0,0 +1,89 @@
+import { Network } from '../types/types';
+
+type NetworkPageKeys = Map<Network, NetworkPageKey>;
+
+interface UnichainPageKeyValue {
+  networkPageKeys: NetworkPageKeys;
+  expiresAt: number;
+}
+
+export class NetworkPageKey {
+  // When a network has no next page we will save a page key with this
+  // value. The hasNextPage() method will return false for any page key
+  // matching this value.
+  private NO_MORE_PAGES = 'NO_MORE_PAGES';
+
+  private pageKey: string;
+
+  constructor(pageKey?: any) {
+    this.pageKey = pageKey ? pageKey.toString() : this.NO_MORE_PAGES;
+  }
+
+  public value(): string {
+    return this.pageKey;
+  }
+
+  public hasNextPage(): boolean {
+    return this.value() !== this.NO_MORE_PAGES;
+  }
+}
+
+// As of today, the API expires page keys in 15 minutes. We set the
+// unichain expiration to be less than that in order to ensure that
+// we never deal with expired network keys.
+const DEFAULT_UNICHAIN_PAGE_KEY_EXPIRATION_MS = 1_000 * 60 * 10;
+
+export class UnichainPageKeyCache {
+  private pageKeyExpirationMs: number;
+  private cache: Record<string, UnichainPageKeyValue>;
+
+  constructor(
+    cache: Record<string, UnichainPageKeyValue> = {},
+    pageKeyExpirationMs: number = DEFAULT_UNICHAIN_PAGE_KEY_EXPIRATION_MS
+  ) {
+    this.cache = cache;
+    this.pageKeyExpirationMs = pageKeyExpirationMs;
+  }
+
+  public static generateKey(networkPageKeys: NetworkPageKeys): string | null {
+    const anyHasNextPage = Array.from(networkPageKeys.values()).find(key =>
+      key.hasNextPage()
+    );
+    if (!anyHasNextPage) {
+      return null;
+    }
+
+    return `${Date.now()}${Math.round(Math.random() * 100000000)}`;
+  }
+
+  public set(unichainKey: string, networkPageKeys: NetworkPageKeys): void {
+    this.cache[unichainKey] = {
+      networkPageKeys,
+      expiresAt: Date.now() + this.pageKeyExpirationMs
+    };
+  }
+
+  public get(unichainKey: string): UnichainPageKeyValue | null {
+    return this.cache[unichainKey] || null;
```

Comment:
This is fine, but I like to use `??` instead of `||` when expressing default values like this. It does the same thing, but only uses the default value if the left-hand-side is `null` or `undefined`, rather than any falsy value.

Likewise in test file.